pkg/v2/tkr/webhook/cluster/tkr-resolver/cluster/cluster.go

```diff
@@ -6,16 +6,335 @@ package cluster
+// Pre-reqs: cluster != nil && clusterClass != nil
+func (cw *Webhook) constructQuery(cluster *clusterv1.Cluster, clusterClass *clusterv1.ClusterClass) (data.Query, error) {
+	tkrSelector, err := selectorFromAnnotation(cluster.Annotations, clusterClass.Annotations, runv1.AnnotationResolveTKR)
+	if tkrSelector == nil || cluster.Spec.Topology == nil {
+		return data.Query{}, err
+	}
+
+	osImageSelector, err := selectorFromAnnotation(
+		cluster.Spec.Topology.ControlPlane.Metadata.Annotations,
+		clusterClass.Spec.ControlPlane.Metadata.Annotations,
+		runv1.AnnotationResolveOSImage)
+	if err != nil {
+		return data.Query{}, err
+	}
+	if osImageSelector == nil {
+		osImageSelector = labels.Everything() // default to empty selector (matches all) for OSImages
+	}
+
+	cpQuery := cw.constructOSImageQuery(cluster.Spec.Topology.Version, tkrSelector, osImageSelector, labels.Merge(cluster.Labels, cluster.Spec.Topology.ControlPlane.Metadata.Labels))
+
+	if cluster.Spec.Topology.Workers == nil {
+		return data.Query{ControlPlane: cpQuery}, nil
+	}
+
+	mdQueries, err := cw.constructMDQueries(cluster, clusterClass, tkrSelector)
+	if err != nil {
+		return data.Query{}, err
+	}
+
+	return data.Query{ControlPlane: cpQuery, MachineDeployments: mdQueries}, nil
+}
+
+// selectorFromAnnotation produces a selector from the value of the specified annotation.
+func selectorFromAnnotation(cAnnots, ccAnnots map[string]string, annotation string) (labels.Selector, error) {
+	var selectorStr *string
+	if selectorStr = getAnnotation(cAnnots, annotation); selectorStr == nil {
+		if selectorStr = getAnnotation(ccAnnots, annotation); selectorStr == nil {
+			return nil, nil
+		}
+	}
+	selector, err := labels.Parse(*selectorStr)
+	return selector, errors.Wrapf(err, "error parsing selector: '%s'", *selectorStr)
+}
+
+// getAnnotation gets the value of the annotation specified by name.
+// Returns nil if such annotation is not found.
+func getAnnotation(annotations map[string]string, name string) *string {
+	if annotations == nil {
+		return nil
+	}
+	value, exists := annotations[name]
+	if !exists {
+		return nil
+	}
+	return &value
+}
+
+func (cw *Webhook) constructOSImageQuery(v string, tkrSelector, osImageSelector labels.Selector, labelSet labels.Set) *data.OSImageQuery {
+	if tkrName, ok := labelSet[runv1.LabelTKR]; ok {
+		if tkr := cw.TKRResolver.Get(tkrName, &runv1.TanzuKubernetesRelease{}).(*runv1.TanzuKubernetesRelease); tkr != nil {
+			if osImageName, ok := labelSet[runv1.LabelOSImage]; ok {
+				if osImage := cw.TKRResolver.Get(osImageName, &runv1.OSImage{}).(*runv1.OSImage); osImage != nil {
+					req, _ := labels.NewRequirement(version.Label(v), selection.Exists, nil)
+					tkrSelectorWithVersion := tkrSelector.Add(*req)
+					osImageSelectorWithVersion := osImageSelector.Add(*req)
+					if tkrSelectorWithVersion.Matches(labels.Set(tkr.Labels)) && osImageSelectorWithVersion.Matches(labels.Set(osImage.Labels)) {
```

Comment:
nit: Can we add comment/notes for this. It needed a bit of effort to understand this.

Reply:
Done